lettuce/ext/_flows/taylorgreen.py

```diff
@@ -1,85 +1,119 @@
-
-    def make_resolution(self, resolution: Union[int, List[int]], stencil: Optional['Stencil'] = None) -> List[int]:
+        if stencil is None and not isinstance(resolution, list):
+            warnings.warn("Requiring information about dimensionality!"
+                          " Either via stencil or resolution. Setting "
+                          "dimension to 2.", UserWarning)
+            self.d = 2
+        else:
+            self.stencil = stencil() if callable(stencil) else stencil
+            self.d = self.stencil.d if self.stencil is not None else len(
+                resolution)
+        ExtFlow.__init__(self, context, resolution, reynolds_number,
+                         mach_number, stencil, equilibrium)
+
+    def make_resolution(self, resolution: Union[int, List[int]],
+                        stencil: Optional['Stencil'] = None) -> List[int]:
         if isinstance(resolution, int):
-            return [resolution] * 2
+            return [resolution] * self.d
         else:
-            assert len(resolution) == 2, 'the resolution of a 2d taylor green vortex must obviously be 2!'
+            assert len(resolution) in [2, 3], ('the resolution of a '
+                                               'taylor-green-vortex '
+                                               'must be 2- or 3-dimensional!')
             return resolution
 
-    def make_units(self, reynolds_number, mach_number, resolution) -> 'UnitConversion':
+    def make_units(self, reynolds_number, mach_number,
+                   resolution) -> 'UnitConversion':
         return UnitConversion(
             reynolds_number=reynolds_number,
             mach_number=mach_number,
-            characteristic_length_lu=resolution[0],
-            characteristic_length_pu=2 * np.pi,
+            characteristic_length_lu=resolution[0] / (2 * torch.pi),
+            characteristic_length_pu=1,
             characteristic_velocity_pu=1)
 
     @property
     def grid(self):
-        x = np.linspace(0, 2 * np.pi, num=self.resolution[0], endpoint=False)
-        y = np.linspace(0, 2 * np.pi, num=self.resolution[1], endpoint=False)
-        return np.meshgrid(x, y, indexing='ij')
-
-    def initial_pu(self) -> (float, float):
+        endpoints = [2 * torch.pi * (1 - 1 / n) for n in
+                     self.resolution]  # like endpoint=False in np.linspace
+        xyz = tuple(torch.linspace(0, endpoints[n],
+                                   steps=self.resolution[n],
+                                   device=self.context.device,
+                                   dtype=self.context.dtype)
+                    for n in range(self.d))
+        return torch.meshgrid(*xyz, indexing='ij')
+
+    def initial_pu(self) -> (torch.Tensor, torch.Tensor):
         return self.analytic_solution(t=0)
 
-    def analytic_solution(self, t=0):
+    def analytic_solution(self, t: float) -> (torch.Tensor, torch.Tensor):
```

Comment:
Add a warning when `stencil.d == 3` and the analytic solution is called since its only true for 2D

Reply:
ah, at least, when the time is `t>0 and stencil.d ==3`

Reply:
see [0297d37](https://github.com/lettucecfd/lettuce/pull/200/commits/0297d37fb2785f9ad8ac347fe3e475a9e89927b1)